src/commands/init.js

```diff
@@ -1,64 +1,203 @@
+        try {
+          abiFromEtherscan = await loadAbiFromEtherscan(network, value)
+        } catch (e) {}
+        return value
+      },
+    },
+    {
+      type: 'input',
+      name: 'abi',
+      message: 'ABI file (path)',
+      skip: () => fromExample !== undefined || abiFromEtherscan !== undefined,
+      validate: async value => {
+        if (fromExample || abiFromEtherscan) {
+          return true
+        }
+
+        let exists = await toolbox.filesystem.exists(value)
+        if (!exists) {
+          return 'File does not exist.'
+        } else if (exists === 'dir') {
+          return 'Path points to a directory, not a file.'
+        } else if (exists === 'other') {
+          return 'Not sure what this path points to.'
+        } else {
+          abiFromFile = await toolbox.filesystem.read(value, 'json')
+          return abiFromFile ? true : 'Not a valid ABI file'
+        }
+      },
+    },
+  ]
+
+  try {
+    let answers = await toolbox.prompt.ask(questions)
+    return { ...answers, abi: abiFromEtherscan || abiFromFile }
+  } catch (e) {
+    return undefined
+  }
+}
+
+const loadAbiFromEtherscan = async (network, address) =>
+  await withSpinner(
+    `Fetching ABI from Etherscan`,
+    `Failed to fetch ABI from Etherscan`,
+    async spinner => {
+      let result = await fetch(
+        `https://${
+          network === 'mainnet' ? 'api' : `api-${network}`
+        }.etherscan.io/api?module=contract&action=getabi&address=${address}`
+      )
+      let json = await result.json()
+      return JSON.parse(json.result)
+    }
+  )
+
 module.exports = {
   description: 'Creates a new subgraph with basic scaffolding',
+  options: {
+    boolean: ['from-example'],
+  },
   run: async toolbox => {
     // Obtain tools
-    let { filesystem, print, system } = toolbox
+    let { print, system } = toolbox
 
     // Read CLI parameters
-    let { allowSimpleName, h, help } = toolbox.parameters.options
-    let subgraphName = toolbox.parameters.first
+    let {
+      abi,
+      allowSimpleName,
+      fromContract,
+      fromExample,
+      h,
+      help,
+      network,
+    } = toolbox.parameters.options
 
-    // Show help text if requested
-    if (help || h) {
-      print.info(HELP)
-      return
-    }
-
-    // Validate the subgraph name
-    if (!subgraphName) {
-      print.error('No subgraph name provided')
-      print.info(HELP)
+    if (fromContract && fromExample) {
+      print.error(`Only one of --from-example and --from-contract can be used at a time.`)
       process.exitCode = 1
       return
     }
-    try {
-      validateSubgraphName(subgraphName, { allowSimpleName })
-    } catch (e) {
-      print.error(e.message)
-      print.info(`
-Examples:
 
-  $ graph init ${os.userInfo().username}/${subgraphName}
-  $ graph init ${subgraphName} --allow-simple-name`)
-      process.exitCode = 1
-      return
-    }
+    let [subgraphName, directory] = fixParameters(toolbox.parameters, {
```

Comment:
I don't understand what `fixParameters` does.

Reply:
Basically, if you run `graph init --from-example leodasvacas/some-subgraph`, you would expect `toolbox.parameters.first` to be `'leodasvacas/some-subgraph'` and `toolbox.parameters.array` to be `['leodasvacas/some-subgraph']`. However, due to the way Gluegun parses command line arguments, it actually considers `leodasvacas/some-subgraph` to be the value for the `--from-example` option, so `fromExample` would be `'leodasvacas/some-subgraph'` and `tools.parameters.first` would be `undefined`. Ouch.

`fixParameters` checks all provided boolean options for whether they have an argument value; if that is the case, it pushes that to the front of the `toolbox.parameters.array` and returns the result. It's a hack to work around the above problem.

I had some fun with this last week: https://github.com/infinitered/gluegun/pull/464

This reminds me though: we should be using `fixParameters` in all our commands, because invocations like `graph deploy --watch leodasvacas/some-subgraph` are also affected. And I should add a comment. 😄

Reply:
Tracked here: https://github.com/graphprotocol/graph-cli/issues/231

Reply:
ah, awesome that you're trying to fix it upstream.

Reply:
Did you get to adding a comment?

Reply:
I actually rewrote the whole function and added a comment:

https://github.com/graphprotocol/graph-cli/blob/1bf7373ad7925c48b16288937fdf6f765ab28061/src/command-helpers/gluegun.js#L1-L47

Reply:
Ah yes sorry. Great comment!